rust/theoros/src/handlers/websocket/subscribe_to_calldata.rs

```diff
@@ -0,0 +1,354 @@
+use {
+    crate::{
+        configs::evm_config::EvmChainName,
+        types::{
+            hyperlane::{CheckpointMatchEvent, DispatchUpdate},
+            pragma::{
+                calldata::{AsCalldata, HyperlaneMessage, Payload},
+                constants::{DEFAULT_ACTIVE_CHAIN, HYPERLANE_VERSION, MAX_CLIENT_MESSAGE_SIZE, PING_INTERVAL_DURATION},
+            },
+            rpc::RpcDataFeed,
+        },
+        AppState,
+    },
+    alloy::hex,
+    anyhow::{anyhow, Result},
+    axum::{
+        extract::{
+            ws::{Message, WebSocket, WebSocketUpgrade},
+            ConnectInfo, State as AxumState,
+        },
+        response::IntoResponse,
+    },
+    futures::{
+        stream::{SplitSink, SplitStream},
+        SinkExt, StreamExt,
+    },
+    serde::{Deserialize, Serialize},
+    std::{
+        collections::HashMap,
+        net::SocketAddr,
+        sync::{
+            atomic::{AtomicUsize, Ordering},
+            Arc,
+        },
+    },
+    tokio::sync::{broadcast::Receiver, watch},
+};
+
+#[derive(Clone)]
+pub struct DataFeedClientConfig {}
+
+pub struct WsState {
+    pub subscriber_counter: AtomicUsize,
+}
+
+#[allow(clippy::new_without_default)]
+impl WsState {
+    pub fn new() -> Self {
+        Self { subscriber_counter: AtomicUsize::new(0) }
+    }
+}
+
+#[derive(Deserialize, Debug, Clone)]
+#[serde(tag = "type")]
+enum ClientMessage {
+    #[serde(rename = "subscribe")]
+    Subscribe { ids: Vec<String>, chain_name: EvmChainName },
+    #[serde(rename = "unsubscribe")]
+    Unsubscribe { ids: Vec<String> },
+}
+
+#[derive(Serialize, Debug, Clone)]
+#[serde(tag = "type")]
+enum ServerMessage {
+    #[serde(rename = "response")]
+    Response(ServerResponseMessage),
+    #[serde(rename = "data_feed_update")]
+    DataFeedUpdate { data_feed: RpcDataFeed },
+}
+
+#[derive(Serialize, Debug, Clone)]
+#[serde(tag = "status")]
+enum ServerResponseMessage {
+    #[serde(rename = "success")]
+    Success,
+    #[serde(rename = "error")]
+    Err { error: String },
+}
+
+pub async fn ws_route_handler(
+    ws: WebSocketUpgrade,
+    AxumState(state): AxumState<AppState>,
+    ConnectInfo(_): ConnectInfo<SocketAddr>,
+) -> impl IntoResponse {
+    ws.max_message_size(MAX_CLIENT_MESSAGE_SIZE).on_upgrade(move |socket| websocket_handler(socket, state))
+}
+
+#[tracing::instrument(skip(stream, state))]
+async fn websocket_handler(stream: WebSocket, state: AppState) {
+    let ws_state = state.ws.clone();
+
+    // TODO: add new connection to metrics
+
+    let (sender, receiver) = stream.split();
+
+    let feeds_receiver = state.storage.feeds_channel.subscribe();
+
+    let id = ws_state.subscriber_counter.fetch_add(1, Ordering::SeqCst);
+
+    let mut subscriber = Subscriber::new(id, Arc::new(state.clone()), feeds_receiver, receiver, sender);
```

Comment:
Is it really needed to clone the inner state? Why don't we just use an Arc?

Reply:
done